includes/API_Credentials/API_Credentials_Manager.php

```
@@ -0,0 +1,270 @@
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array<string, ProviderMetadata> Array of cloud provider metadata objects, keyed by provider ID.
+	 */
+	private function get_all_cloud_providers_metadata(): array {
+		$all_providers = $this->get_all_providers_metadata();
+
+		return array_filter(
+			$all_providers,
+			function ( ProviderMetadata $metadata ) {
+				return $metadata->getType()->isCloud();
+			}
+		);
+	}
+
+	/**
+	 * Registers the settings for storing the API credentials.
+	 *
+	 * The setting will only be registered once, even if the class is used multiple times.
+	 *
+	 * @since n.e.x.t
+	 */
+	private function register_settings(): void {
+		// Avoid registering the setting multiple times.
+		$registered_settings = get_registered_settings();
+		if ( isset( $registered_settings[ self::OPTION_PROVIDER_CREDENTIALS ] ) ) {
+			return;
+		}
+
+		register_setting(
+			self::OPTION_GROUP,
+			self::OPTION_PROVIDER_CREDENTIALS,
+			array(
+				'type'              => 'object',
+				'default'           => array(),
+				'sanitize_callback' => function ( $credentials ) {
+					if ( ! is_array( $credentials ) ) {
+						return array();
+					}
+
+					// Assume that all cloud providers require an API key.
+					$providers_metadata_keyed_by_ids = $this->get_all_cloud_providers_metadata();
+
+					$credentials = array_intersect_key( $credentials, $providers_metadata_keyed_by_ids );
+					foreach ( $credentials as $provider_id => $api_key ) {
+						$credentials[ $provider_id ] = sanitize_text_field( $api_key );
+					}
+					return $credentials;
+				},
+			)
+		);
+	}
+
+	/**
+	 * Passes the stored API credentials to the PHP AI Client SDK.
+	 *
+	 * This method should be called on every request, before any API requests are made via the PHP AI Client SDK.
+	 *
+	 * @since n.e.x.t
+	 */
+	private function pass_credentials_to_client(): void {
+		$credentials = get_option( self::OPTION_PROVIDER_CREDENTIALS, array() );
+		if ( ! is_array( $credentials ) ) {
+			return;
```

Feels like we should have a `_doing_it_wrong()` or something here, because this should never happen.

---

Maybe. I think the `_doing_it_wrong()` should happen at the point where some invalid value would be stored in the option. To me, here it feels out of place since the flaw would likely have occurred elsewhere, and this is called on basically every request?

---

Yeah, you're right. I'm just leery of silent failures, as it would be more difficult for someone later to figure out that the providers aren't working properly because the options somehow got messed up. I'd probably throw a `RuntimeException` here, as clearly something is very wrong to have gotten to this point. To clear, "this point" being that the option isn't even empty, it's an unexpected value entirely.

---

Improved in https://github.com/WordPress/wp-ai-client/pull/6/commits/9091409987573b1a2f5271fe2ec2fc35bfa9c805